README
NAME
    Encode::IBM - IBM-specific encoding mappings

VERSION
    This document describes version 0.06 of Encode::IBM, released September
    25, 2007.

SYNOPSIS
        use Encode;
        use Encode::IBM;
        binmode( STDIN, ':encoding(ibm-835)' );
        binmode( STDOUT, ':encoding(ibm-947-sosi)' );

DESCRIPTIONS
    This module provides encoding extensions of IBM-specific mappings. The
    sources are from the ICU mapping repository:

        http://oss.software.ibm.com/cvs/icu/charset/data/ucm/

ENCODINGS
        ibm-835
        ibm-937
        ibm-939
        ibm-947

        ibm-835-sosi
        ibm-939-sosi
        ibm-947-sosi

SEE ALSO
    Encode

COPYRIGHT
    Copyright 2004, 2005, 2006, 2007 by Audrey Tang <cpan@audreyt.org>.

    This software is released under the MIT license cited below.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
[Changes for 0.10 - 2008-01-07]

* Added ibm-964 encoding for EUC-TW.

* Upgraded Makfile.PL to latest enc2xs. (No user-visible changes.)

[Changes for 0.07 - 2007-11-09]

* Fix Win32 building problems caused by XSLoader not functioning.
  Reported by: Sam Chen

* Fix Makefile.PL's issue with earlier MakeMaker without SIGN=>1 support.
  Reported by: Sam Chen

[Changes for 0.06 - 2007-09-25]

* LICENSING CHANGE: This compilation and all individual files in it
  are now under the permissive "MIT" license.  See the COPYRIGHT
  section in README for the new terms.

* Added ibm-939 and ibm-939-sosi encoding.

* Added ibm-835-sosi-tsgh, a variant of ibm-835-sosi with 
  0x28/0x29 instead of 0x0E/0x0F as shift codes, plus a few
  control codes for font sizing.

[Changes for 0.05 - 2006-09-28]

* Add ibm-835-sosi encoding.

* Run "enc2xs -C" on "make install" by default.

[Changes for 0.04 - 2006-09-15]

* Add ibm-937 and ibm-947-sosi encodings.

[Changes for 0.02 - 2004-12-30]

* First really working CPAN release.


lib/Encode/IBM.pm
package Encode::IBM;
 
use 5.008;
use strict;
use Encode;
use Encode::IBM::835SOSI;
use Encode::IBM::835SOSI::TSGH;
use Encode::IBM::947SOSI;
use Encode::IBM::939SOSI;

BEGIN {
    our $VERSION = '0.10';

    local $@;
    eval {
        require XSLoader;
        XSLoader::load(__PACKAGE__, $VERSION);
        1;
    } or do {
        require DynaLoader;
        push our @ISA, 'DynaLoader';
        __PACKAGE__->bootstrap($VERSION);
    };
};

1;

__END__

=head1 NAME
 
Encode::IBM - IBM-specific encoding mappings

=head1 VERSION

This document describes version 0.10 of Encode::IBM, released
January 7, 2008.

=head1 SYNOPSIS

    use Encode;
    use Encode::IBM;
    binmode( STDIN, ':encoding(ibm-835)' );
    binmode( STDOUT, ':encoding(ibm-947-sosi)' );

=head1 DESCRIPTIONS

This module provides encoding extensions of IBM-specific mappings.
The sources are from the ICU mapping repository:

    http://oss.software.ibm.com/cvs/icu/charset/data/ucm/

=head1 ENCODINGS

    ibm-835
    ibm-937
    ibm-939
    ibm-947

    ibm-835-sosi
    ibm-939-sosi
    ibm-947-sosi

=head1 SEE ALSO

L<Encode>

=head1 COPYRIGHT

Copyright 2004, 2005, 2006, 2007, 2008 by Audrey Tang E<lt>cpan@audreyt.orgE<gt>.

This software is released under the MIT license cited below.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut


lib/Encode/IBM/835SOSI.pm
package Encode::IBM::835SOSI;

use strict;
use vars qw($VERSION);
$VERSION = '0.10';

use Encode ();

use base qw(Encode::Encoding);
__PACKAGE__->Define('ibm-835-sosi');

my $base37;
my $base835;

sub decode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base835 ||= Encode::find_encoding('ibm-835');

    my $out;
    foreach my $chunk (split(/\x0E([^\x0F]*\x0F)/, $str)) {
        if ($chunk =~ /\x0F\z/) {
            chop $chunk;
            $out .= $base835->decode($chunk);
        }
        else {
            $out .= $base37->decode($chunk);
        }
    }
    return $out;
}

sub encode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base835 ||= Encode::find_encoding('ibm-835');

    if ($str =~ s/^([\x00-\xff]+)//) {
        # english
        my $sub = $1;
        return $base37->encode($sub) . $obj->encode($str, $chk);
    }
    elsif ($str =~ s/^([^\x00-\xff]+)//) {
        # chinese - shift in + shift out
        my $sub = $1;
        return "\x0E".$base835->encode($sub)."\x0F".$obj->encode($str, $chk);
    }
}

1;
__END__


=head1 NAME

Encode::IBM::835SOSI

=cut


lib/Encode/IBM/939SOSI.pm
package Encode::IBM::939SOSI;

use strict;
use vars qw($VERSION);
$VERSION = '0.10';

use Encode ();

use base qw(Encode::Encoding);
__PACKAGE__->Define('ibm-939-sosi');

my $base37;
my $base939;

sub decode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base939 ||= Encode::find_encoding('ibm-939');

    my $out;
    foreach my $chunk (split(/\x0E([^\x0F]*\x0F)/, $str)) {
        if ($chunk =~ /\x0F\z/) {
            chop $chunk;
            $out .= $base939->decode($chunk);
        }
        else {
            $out .= $base37->decode($chunk);
        }
    }
    return $out;
}

sub encode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base939 ||= Encode::find_encoding('ibm-939');

    if ($str =~ s/^([\x00-\xff]+)//) {
        # english
        my $sub = $1;
        return $base37->encode($sub) . $obj->encode($str, $chk);
    }
    elsif ($str =~ s/^([^\x00-\xff]+)//) {
        # chinese - shift in + shift out
        my $sub = $1;
        return "\x0E".$base939->encode($sub)."\x0F".$obj->encode($str, $chk);
    }
}

1;
__END__


=head1 NAME

Encode::IBM::939SOSI

=cut

=cut


lib/Encode/IBM/947SOSI.pm
package Encode::IBM::947SOSI;

use strict;
use vars qw($VERSION);
$VERSION = '0.10';

use Encode ();

use base qw(Encode::Encoding);
__PACKAGE__->Define('ibm-947-sosi');

my $base37;
my $base947;

sub decode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base947 ||= Encode::find_encoding('ibm-947');

    my $out;
    foreach my $chunk (split(/\x0E([^\x0F]*\x0F)/, $str)) {
        if ($chunk =~ /\x0F\z/) {
            chop $chunk;
            $out .= $base947->decode($chunk);
        }
        else {
            $out .= $base37->decode($chunk);
        }
    }
    return $out;
}

sub encode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base947 ||= Encode::find_encoding('ibm-947');

    if ($str =~ s/^([\x00-\xff]+)//) {
        # english
        my $sub = $1;
        return $base37->encode($sub) . $obj->encode($str, $chk);
    }
    elsif ($str =~ s/^([^\x00-\xff]+)//) {
        # chinese - shift in + shift out
        my $sub = $1;
        return "\x0E".$base947->encode($sub)."\x0F".$obj->encode($str, $chk);
    }
}

1;
__END__


=head1 NAME

Encode::IBM::947SOSI

=cut

=cut


lib/Encode/IBM/835SOSI/TSGH.pm
package Encode::IBM::835SOSI::TSGH;

use strict;
use vars qw($VERSION);
$VERSION = '0.10';

use Encode ();

use base qw(Encode::Encoding);
__PACKAGE__->Define('ibm-835-sosi-tsgh');

my $base37;
my $base835;

sub decode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base835 ||= Encode::find_encoding('ibm-835');
    #$str =~ s/\x00/K/g;
    $str =~ s/\x00/\x40/g;
    $str =~ s/\x30\xe1/Fg/g;    # MAKE SMALL
    $str =~ s/\x30\xe0//g;      # MAKE LARGE
    #$str =~ s/\x30\xe0/Mh/g;

    my $out;
    foreach my $chunk (split(/\x28([^\x29]*\x29)/, $str)) {
        if ($chunk =~ /\x29\z/) {
            chop $chunk;
            $chunk =~ s/\@\@/iJ/g;
            $out .= $base835->decode($chunk);
        }
        else {
            $out .= $base37->decode($chunk);
        }
    }
    return $out;
}

sub encode
{
    my ($obj,$str,$chk) = @_;

    $base37 ||= Encode::find_encoding('cp37');
    $base835 ||= Encode::find_encoding('ibm-835');

    if ($str =~ s/^([\x00-\xff]+)//) {
        # english
        my $sub = $1;
        return $base37->encode($sub) . $obj->encode($str, $chk);
    }
    elsif ($str =~ s/^([^\x00-\xff]+)//) {
        # chinese - shift in + shift out
        my $sub = $1;
        return "\x28".$base835->encode($sub)."\x29".$obj->encode($str, $chk);
    }
}

1;
__END__


=head1 NAME

Encode::IBM::835SOSI::TSGH

=cut


Makefile.PL
#
# This file is auto-generated by:
# enc2xs version 1.32
# Thu Sep 16 16:05:12 2004
#
use 5.7.2;
use strict;
use ExtUtils::MakeMaker;
use Config;

# Please edit the following to the taste!
my $name = 'IBM';
my %tables = (
	     IBM_t   => [ 'ucm/ibm-947.ucm','ucm/ibm-835.ucm','ucm/ibm-937.ucm', 'ucm/ibm-939.ucm' ],
	     );

#### DO NOT EDIT BEYOND THIS POINT!
require File::Spec;
my ($enc2xs, $encode_h) = ();
PATHLOOP:
for my $d (@Config{qw/bin sitebin vendorbin/}, 
	   (split /$Config{path_sep}/o, $ENV{PATH})){
    for my $f (map {$_, "$_.bat"} ('enc2xs', "enc2xs$Config{version}")) {
        my $path = File::Spec->catfile($d, $f);
        -r $path and $enc2xs = $path and last PATHLOOP;
    }
}
$enc2xs or die "enc2xs not found!";
print "enc2xs is $enc2xs\n";
my %encode_h = ();
for my $d (@INC){
    my $dir = File::Spec->catfile($d, "Encode");
    my $file =  File::Spec->catfile($dir, "encode.h");
    -f $file and $encode_h{$dir} = -M $file;
}
%encode_h or die "encode.h not found!";
# find the latest one
($encode_h) = sort {$encode_h{$b} <=> $encode_h{$a}} keys %encode_h;
print "encode.h is at $encode_h\n";

WriteMakefile(
              INC		=> "-I$encode_h",
	      NAME		=> 'Encode::'.$name,
	      ABSTRACT		=> 'IBM-specific encoding mappings',
	      AUTHOR		=> 'Audrey Tang <cpan@audreyt.org>',
	      VERSION_FROM	=> "lib/Encode/$name.pm",
	      OBJECT		=> '$(O_FILES)',
              SIGN              => (do { local $@; eval { ExtUtils::MakeMaker->VERSION(6.31) } } || 0),
	      'dist'		=> {
		  COMPRESS	=> 'gzip -9f',
		  SUFFIX	=> 'gz',
		  DIST_DEFAULT => 'all tardist',
	      },
	      MAN3PODS	=> {},
	      PREREQ_PM => {
			    'Encode'     => "1.41",
                           },
	      # OS 390 winges about line numbers > 64K ???
	      XSOPT => '-nolinenumbers',
	      );

# add 'enc2xs -C' to the 'install' target
{
    local $/;
    open MAKEFILE, 'Makefile' or die "Can't read Makefile: $!";
    my $makefile = <MAKEFILE>;
    close MAKEFILE;
    $makefile =~ s/(install :: .*)/$1\n\t\@$enc2xs -C/;
    open MAKEFILE, '>Makefile' or die "Can't write to Makefile: $!";
    print MAKEFILE $makefile;
    close MAKEFILE;
}
package MY;

sub post_initialize
{
    my ($self) = @_;
    my %o;
    my $x = $self->{'OBJ_EXT'};
    # Add the table O_FILES
    foreach my $e (keys %tables)
    {
	$o{$e.$x} = 1;
    }
    $o{"$name$x"} = 1;
    $self->{'O_FILES'} = [sort keys %o];
    my @files = ("$name.xs");
    $self->{'C'} = ["$name.c"];
    # The next two lines to make MacPerl Happy -- dankogai via pudge
    $self->{SOURCE} .= " $name.c"
        if $^O eq 'MacOS' && $self->{SOURCE} !~ /\b$name\.c\b/;
    # $self->{'H'} = [$self->catfile($self->updir,'encode.h')];
    my %xs;
    foreach my $table (keys %tables) {
	push (@{$self->{'C'}},"$table.c");
	# Do NOT add $table.h etc. to H_FILES unless we own up as to how they
	# get built.
	foreach my $ext (qw($(OBJ_EXT) .c .h .exh .fnm)) {
	    push (@files,$table.$ext);
	}
    }
    $self->{'XS'} = { "$name.xs" => "$name.c" };
    $self->{'clean'}{'FILES'} .= join(' ',@files);
    open(XS,">$name.xs") || die "Cannot open $name.xs:$!";
    print XS <<'END';
#include <EXTERN.h>
#include <perl.h>
#include <XSUB.h>
#define U8 U8
#include "encode.h"
END
    foreach my $table (keys %tables) {
	print XS qq[#include "${table}.h"\n];
    }
    print XS <<"END";

static void
Encode_XSEncoding(pTHX_ encode_t *enc)
{
 dSP;
 HV *stash = gv_stashpv("Encode::XS", TRUE);
 SV *sv    = sv_bless(newRV_noinc(newSViv(PTR2IV(enc))),stash);
 int i = 0;
 PUSHMARK(sp);
 XPUSHs(sv);
 while (enc->name[i])
  {
   const char *name = enc->name[i++];
   XPUSHs(sv_2mortal(newSVpvn(name,strlen(name))));
  }
 PUTBACK;
 call_pv("Encode::define_encoding",G_DISCARD);
 SvREFCNT_dec(sv);
}

MODULE = Encode::$name	PACKAGE = Encode::$name
PROTOTYPES: DISABLE
BOOT:
{
END
    foreach my $table (keys %tables) {
	print XS qq[#include "${table}.exh"\n];
    }
    print XS "}\n";
    close(XS);
    return "# Built $name.xs\n\n";
}

sub postamble
{
    my $self = shift;
    my $dir  = "."; # $self->catdir('Encode');
    my $str  = "# $name\$(OBJ_EXT) depends on .h and .exh files not .c files - but all written by enc2xs\n";
    $str    .= "$name.c : $name.xs ";
    foreach my $table (keys %tables)
    {
	$str .= " $table.c";
    }
    $str .= "\n\n";
    $str .= "$name\$(OBJ_EXT) : $name.c\n\n";

    foreach my $table (keys %tables)
    {
	my $numlines = 1;
	my $lengthsofar = length($str);
	my $continuator = '';
	$str .= "$table.c : Makefile.PL";
	foreach my $file (@{$tables{$table}})
	{
	    $str .= $continuator.' '.$self->catfile($dir,$file);
	    if ( length($str)-$lengthsofar > 128*$numlines )
	    {
		$continuator .= " \\\n\t";
		$numlines++;
	    } else {
		$continuator = '';
	    }
	}
	my $plib   = $self->{PERL_CORE} ? '"-I$(PERL_LIB)"' : '';
	my $ucopts = '-"Q"';
	$str .=  
	    qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
	open (FILELIST, ">$table.fnm")
	    || die "Could not open $table.fnm: $!";
	foreach my $file (@{$tables{$table}})
	{
	    print FILELIST $self->catfile($dir,$file) . "\n";
	}
	close(FILELIST);
    }
    return $str;
}

